File: 208-implement-trie.ts
class TrieNode {
  children: (TrieNode | undefined)[] = Array(26)
  isTerminator: boolean = false
}

const CHAR_INDEX: { [index: string]: number } = "abcdefghijklmnopqrstuvwxyz"
  .split("")
  .reduce((acc, char, i) => ({ ...acc, [char]: i }), {})

class Trie {
  children: (TrieNode | undefined)[] = Array(26)

  constructor() {}

  insert(word: string): void {
    let { children } = this
    for (let i = 0; i < word.length; i++) {
      let char = word[i]
      const charIndex = CHAR_INDEX[char]
      const isLast = i === word.length - 1

      if (children[charIndex] === undefined) {
        children[charIndex] = new TrieNode()
      }

      if (isLast) {
        children[charIndex]!.isTerminator = true
      }

      children = children[charIndex]!.children
    }
  }

  search(word: string): boolean {
    let { children } = this
    for (let i = 0; i < word.length; i++) {
      let char = word[i]
      const charIndex = CHAR_INDEX[char]
      const isLast = i === word.length - 1

      if (children[charIndex] === undefined) {
        return false
      }

      if (isLast) {
        return children[charIndex]!.isTerminator
      }

      children = children[charIndex]!.children
    }

    throw new Error("Invalid input.")
  }

  startsWith(prefix: string): boolean {
    let { children } = this
    for (const char of prefix) {
      const charIndex = CHAR_INDEX[char]
      if (children[charIndex] === undefined) {
        return false
      }
      children = children[charIndex]!.children
    }

    return true
  }
}

/**
 * Your Trie object will be instantiated and called as such:
 * var obj = new Trie()
 * obj.insert(word)
 * var param_2 = obj.search(word)
 * var param_3 = obj.startsWith(prefix)
 */

File: problems/208_implement_trie.py
from __future__ import annotations


class TrieNode:
    children: list[TrieNode | None]
    isEnd: bool

    def __init__(self):
        self.children = [None] * 26
        self.isEnd = False


class Trie:
    root: TrieNode

    def __init__(self) -> None:
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        node = self.root
        for char in word:
            index = ord(char) - ord("a")
            if node.children[index] is None:
                node.children[index] = TrieNode()
            node = node.children[index]
        node.isEnd = True

    def search(self, word: str) -> bool:
        node = self.root
        for char in word:
            index = ord(char) - ord("a")
            if node.children[index] is None:
                return False
            node = node.children[index]
        return node.isEnd

    def startsWith(self, prefix: str) -> bool:
        node = self.root
        for char in prefix:
            index = ord(char) - ord("a")
            if node.children[index] is None:
                return False
            node = node.children[index]
        return True
